### ISSUE TITLE
[SEC] json-rpc server - nbc node crashes on validator_api requests

### ISSUE THREAD

**Issue by tintinweb** [status-im/nimbus-eth2#1680]
## Description

Providing invalid arguments to the rpc methods listed in this issue provoke an nbc node crash. 

### get_v1_beacon_states_stateId_validators - crash in `parsePubkey`

//same for: `get_v1_beacon_states_stateId_validators_validatorId`, `post_v1_validator_duties_attester`

crashes, when provided with a validator pubkey that is of incorrect length. lack of input validation.

```
POST / HTTP/1.1
Host: localhost:9090
Content-Type: application/json
Content-Length: 113
Accept: x/json

{
"jsonrpc":"2.0",
"method":"get_v1_beacon_states_stateId_validators",
"params":["",["0x"],""],
"id":1
}
```

```
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-testutils/testutils/moduletests.nim(21) beacon_node
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(1185) main
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(898) start
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(849) run
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-chronos/chronos/asyncloop.nim(278) poll
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/servers/httpserver.nim(180) processClient
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/server.nim(24) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(166) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(132) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(246) getv1beaconstatesstateIdvalidators
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/validator_api.nim(214) getv1beaconstatesstateIdvalidatorsDoMain
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/validator_api.nim(69) getValidatorInfoFromValidatorId
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/validator_api.nim(37) parsePubkey
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/spec/crypto.nim(266) fromHex
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/spec/crypto.nim(254) fromRaw
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-stew/stew/objects.nim(24) toArray
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nimbus-build-system/vendor/Nim/lib/system/assertions.nim(29) failedAssertImpl
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nimbus-build-system/vendor/Nim/lib/system/assertions.nim(22) raiseAssert
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(49) sysFatal
[[reraised from:
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-testutils/testutils/moduletests.nim(21) beacon_node
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(1185) main
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(898) start
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(849) run
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-chronos/chronos/asyncloop.nim(278) poll
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/servers/httpserver.nim(180) processClient
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/server.nim(24) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(166) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(126) route
]]
Error: unhandled exception: /Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-stew/stew/objects.nim(24, 12) `data.len == N`  [AssertionError]
```

https://github.com/status-im/nim-beacon-chain/blob/c76305f824fbc6a61eca5bbb9d21002eb6c515e3/beacon_chain/validator_api.nim#L36-L40

https://github.com/status-im/nim-beacon-chain/blob/c76305f824fbc6a61eca5bbb9d21002eb6c515e3/beacon_chain/spec/crypto.nim#L250-L255

Which lands in nim-stew. The assertion actually prevents an out of bounds read mem read through `copyMem`.

https://github.com/status-im/nim-stew/blob/152eb1b58cd618a175dc2ae6fba39e620115c356/stew/objects.nim#L23-L25

### get_v1_beacon_states_stateId_validators - crash due to wrong bounds check for validator Idx.

//same for: `get_v1_beacon_states_stateId_validators_validatorId`

The bounds check in `getValidatorInfoFromValidatorId` checks `state.validators.lenu64 >= valIdx`, while it should check for `valIdx >=0 && valIdx < state.validators.lenu64`.

https://github.com/status-im/nim-beacon-chain/blob/c76305f824fbc6a61eca5bbb9d21002eb6c515e3/beacon_chain/validator_api.nim#L77-L79

request: in this example, `len(state.validators)` is about 680. requesting idx 700  causes an `IndexError`.
```
{
"jsonrpc":"2.0",
"method":"get_v1_beacon_states_stateId_validators",
"params":["",["700"],""],
"id":1
}
```

```
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-testutils/testutils/moduletests.nim(21) beacon_node
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(1185) main
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(898) start
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(849) run
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-chronos/chronos/asyncloop.nim(278) poll
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/servers/httpserver.nim(180) processClient
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/server.nim(24) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(166) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(132) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(246) getv1beaconstatesstateIdvalidators
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/validator_api.nim(215) getv1beaconstatesstateIdvalidatorsDoMain
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/spec/datatypes.nim(621) getValidatorInfoFromValidatorId
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(49) sysFatal
[[reraised from:
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-testutils/testutils/moduletests.nim(21) beacon_node
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(1185) main
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(898) start
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(849) run
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-chronos/chronos/asyncloop.nim(278) poll
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/servers/httpserver.nim(180) processClient
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/server.nim(24) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(166) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(126) route
]]
Error: unhandled exception: index 700 not in 0 .. 684 [IndexError]
```

#### get_v1_beacon_states_stateId_committees_epoch - crashes with`index` out of bounds

https://github.com/status-im/nim-beacon-chain/blob/c76305f824fbc6a61eca5bbb9d21002eb6c515e3/beacon_chain/validator_api.nim#L236-L243

request:
```
{
"jsonrpc":"2.0",
"method":"get_v1_beacon_states_stateId_committees_epoch",
"params":["0",0,-1,0,],
"id":1
}
```

```
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-testutils/testutils/moduletests.nim(21) beacon_node
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(1185) main
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(898) start
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(849) run
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-chronos/chronos/asyncloop.nim(278) poll
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/servers/httpserver.nim(180) processClient
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/server.nim(24) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(166) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(132) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(246) getv1beaconstatesstateIdcommitteesepoch
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/validator_api.nim(249) getv1beaconstatesstateIdcommitteesepochDoMain
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/validator_api.nim(245) forSlot
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/validator_api.nim(235) getCommittee
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/spec/validator.nim(234) get_beacon_committee
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/spec/validator.nim(214) compute_committee
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nimbus-build-system/vendor/Nim/lib/system.nim(2494) []
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(49) sysFatal
[[reraised from:
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-testutils/testutils/moduletests.nim(21) beacon_node
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(1185) main
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(898) start
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(849) run
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-chronos/chronos/asyncloop.nim(278) poll
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/servers/httpserver.nim(180) processClient
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/server.nim(24) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(166) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(126) route
]]
Error: unhandled exception: value out of range: -576460752303423466 notin 0 .. 9223372036854775807 [RangeError]
```

### get_v1_beacon_headers_blockId - is not functional at all

```
{
"jsonrpc":"2.0",
"method":"get_v1_beacon_headers_blockId",
"params":["",],
"id":1
}
```

```
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-testutils/testutils/moduletests.nim(21) beacon_node
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(1185) main
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(898) start
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(849) run
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-chronos/chronos/asyncloop.nim(278) poll
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/servers/httpserver.nim(180) processClient
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/server.nim(24) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(166) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(132) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(246) getv1beaconheadersblockId
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/validator_api.nim(267) getv1beaconheadersblockIdDoMain
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(49) sysFatal
[[reraised from:
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-testutils/testutils/moduletests.nim(21) beacon_node
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(1185) main
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(898) start
/Users/tintin/workspace/nim/nim-beacon-chain-round3/beacon_chain/beacon_node.nim(849) run
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-chronos/chronos/asyncloop.nim(278) poll
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/servers/httpserver.nim(180) processClient
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/server.nim(24) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nim-json-rpc/json_rpc/router.nim(166) route
/Users/tintin/workspace/nim/nim-beacon-chain-round3/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(126) route
]]
Error: unhandled exception: 'blob' is not accessible using discriminant 'kind' of type 'BlsValue' [FieldError]
```

## Exploit Scenario

see #1653 

## Mitigation Recommendation

see #1653 

- input validation
- testing, fuzz-testing

**Comment by tintinweb** [comment]
The rpc modules (client/server beacon & validator) lack proper input validation and appear to not be well tested as even some valid requests cause crashes. I recommend to generally review these modules.

**Comment by onqtam** [comment]
This is the associated PR: https://github.com/status-im/nimbus-eth2/pull/1858

**Comment by tintinweb** [comment]
the specific crashes have been resolved with #1858.

* it is recommended to add the failure cases outline here as unit-tests to the project
* generally improving input validation and unit-test coverage (also for known and old failure cases) is advised
> The rpc modules (client/server beacon & validator) lack proper input validation and appear to not be well tested as even some valid requests cause crashes. I recommend to generally review these modules.

* consider fuzz-testing the API
